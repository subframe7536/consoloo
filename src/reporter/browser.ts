import type { LogLevel, Reporter } from '../type'

const scopeColors = ['#3f6894', '#feecd8'] as const
const timeColor = '#918abc'

const levelColors = {
  debug: '#66a2cc',
  info: '#7cbd75',
  warn: '#dbaf57',
  error: '#e08585',
} as const

const r = '.3rem'

function renderBadge(bg: string, fg: string, radius = r): string {
  return `font-size:.8rem;padding:.1rem .3rem;border-radius:${radius};background-color:${bg};color:${fg}`
}

function onBrowserLog(
  date: string,
  msg: any,
  level: LogLevel,
  scope?: string,
  e?: unknown,
): void {
  let _msg = `%c${date} %c${level.toUpperCase()}`
  const args = ['color:' + timeColor]
  if (scope) {
    _msg += `%c${scope}`
    args.push(
      renderBadge(levelColors[level], '#fff', `${r} 0 0 ${r}`),
      renderBadge(scopeColors[0], scopeColors[1], `0 ${r} ${r} 0`),
    )
  } else {
    args.push(renderBadge(levelColors[level], '#fff'))
  }
  _msg += '%c '
  args.push('')
  if (typeof msg !== 'object') {
    _msg += msg
  } else {
    _msg += '%o'
    args.push(msg)
  }
  console.log(_msg, ...args)
  e && console.error(e)
}

function onBrowserTimer(date: string, msg: string, label: string | undefined): void {
  console.log(
    `%c${date} %c${label}%c ${msg}`,
    'color:' + timeColor,
    renderBadge(scopeColors[0], scopeColors[1]),
    '',
  )
}

/**
 * Creates a reporter function that outputs logs to the browser console.
 *
 * Embeded in `createBrowserLogger()`
 *
 * @param timeFormat - A function that formats a Date object into a string timestamp,
 * default is `toLocaleString()`
 *
 * @example
 * ```ts
 * const reporter = createBrowserReporter((date) => date.toISOString());
 * ```
 */
export function createBrowserReporter<T extends string>(
  timeFormat: (date: Date) => string,
): Reporter<T> {
  return (date, msg, level, scope, e) => {
    const d = timeFormat(date)
    level === 'timer'
      ? onBrowserTimer(d, msg, scope)
      : onBrowserLog(d, msg, level, scope, e)
  }
}
