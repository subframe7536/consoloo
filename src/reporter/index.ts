export * from './browser'
export * from './file'
export * from './stdio'
export * from './stream'
