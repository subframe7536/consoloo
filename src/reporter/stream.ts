import type { Reporter } from '../type'
import type { Writable } from 'node:stream'

export interface StreamReporterOptions {
  /**
   * Log time format function
   */
  timeFormat?: (date: Date) => string
  /**
   * Flush queue when it reaches this count
   * @default 100
   */
  flushCount?: number
  /**
   * Auto flush timeout in ms
   * @default 1000
   */
  flushTimeout?: number
  /**
   * Writable stream for log output
   */
  stream: Writable
}

/**
 * Creates a stream-based reporter for logging messages to a writable stream.
 *
 * @param options - {@link StreamReporterOptions Configuration} for the stream reporter
 *
 */
export function createStreamReporter<T extends string>(
  options: StreamReporterOptions,
): Reporter<T> {
  const {
    timeFormat = (date: Date) => date.toLocaleString(),
    flushCount = 100,
    flushTimeout = 1000,
    stream,
  } = options

  const queue: string[] = []
  let timeout: NodeJS.Timeout | null = null

  function clearTimeout(): void {
    if (timeout) {
      globalThis.clearTimeout(timeout)
      timeout = null
    }
  }

  async function flush(): Promise<void> {
    clearTimeout()
    if (stream.writable) {
      stream.write(queue.join(''))
      // clear queue
      queue.length = 0
    }
  }

  return (date, msg, level, scope, e) => {
    const timestamp = timeFormat(date)
    const list = [timestamp, level.toUpperCase()]
    if (scope) {
      list.push(scope)
    }
    list.push(msg)
    let logMessage = list.join(' | ')
    if (e) {
      logMessage += `\n${e.message}\n${e.stack}`
    }

    queue.push(`${logMessage}\n`)

    if (queue.length >= flushCount) {
      flush()
    } else if (!timeout) {
      timeout = setTimeout(() => {
        flush()
      }, flushTimeout)
    }
  }
}
