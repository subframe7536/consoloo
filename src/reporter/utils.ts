import type { LogLevel, ReportFn } from '../type'

export type MessageFormatter<S extends string = string> = ReportFn<string, S>

import { blue, bold, cyan, dim, green, magenta, red, yellow } from 'ansis'
import { isError, toNormalizedError } from 'normal-error'

const colors = {
  info: green,
  debug: blue,
  warn: yellow,
  error: red,
  time: magenta,
}

const CWD_REGEXP = new RegExp(process.cwd().replace(/\\/g, '/'), 'i')
const LOCATION_REGEX = /:[^)]+/
const PAREN_REGEX = /\((.*)\)/
const BACK_SLASH_REGEX = /\\/g
const TAG_REGEX = /(.*): (.*)/
function parseStack(stack: string, level: LogLevel): string {
  const _s = stack.split(/\r?\n/)
  const _stack = _s
    .slice(1)
    .map(l => l
      .replace('file://', '')
      .replace(' at', blue`@`)
      .replace(BACK_SLASH_REGEX, '/')
      .replace(CWD_REGEXP, '.')
      .replace(LOCATION_REGEX, green)
      .replace(PAREN_REGEX, (_, path) => `(${yellow(path)})`),
    )
  return [_s[0].replace(
    TAG_REGEX,
    (_, name: string, msg: string) => `${colors[level].bold(name.toUpperCase())}: ${bold(msg)}`,
  )]
    .concat(_stack)
    .join('\n')
}

function parseMsg(msg: any): string {
  try {
    return typeof msg === 'string'
      ? msg
      : isError(msg)
        ? msg.message
        : JSON.stringify(msg)
  } catch {
    return '' + msg
  }
}

export function defaultStdFormatter<T extends string>(
  date: Date,
  msg: string,
  level: LogLevel | 'timer',
  scope?: T,
  e?: Error,
  timeFormat: (date: Date) => string = d => d.toLocaleString(),
): string {
  if (level === 'timer') {
    return `${colors.time(timeFormat(date))} | ${bold.bgCyan(` ${scope} `)} ${msg}`
  }
  let terminalLog = [
    magenta(date.toLocaleString()),
    colors[level as LogLevel](level.toUpperCase().padEnd(5)),
    (scope ? cyan(scope.padEnd(7)) : dim`default`),
    parseMsg(msg),
  ].join(' | ')

  if (e) {
    const _e = toNormalizedError(e)
    terminalLog += '\n' + parseStack(_e.stack, level)
  }
  return terminalLog
}

export function defaultMessageFormatter<T extends string>(
  date: Date,
  msg: string,
  level: LogLevel | 'timer',
  scope?: T,
  e?: Error,
  timeFormat: (date: Date) => string = d => d.toLocaleString(),
): string {
  const list = [timeFormat(date), level.toUpperCase()]
  if (scope) {
    list.push(scope)
  }
  list.push(msg)
  let logMessage = list.join(' | ')
  if (e) {
    logMessage += `\n${e.message}\n${e.stack}`
  }
  return logMessage
}
