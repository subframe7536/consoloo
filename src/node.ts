import type { Keys, Logger, LoggerOption, LogLevel, LogScope } from './type'
import type { Arrayable } from '@subframe7536/type-utils'
import type { NormalizedError } from 'normal-error'

import { EOL } from 'node:os'

import ansis, { blue, bold, cyan, dim, green, magenta, red, yellow } from 'ansis'
import { isError, toNormalizedError } from 'normal-error'

import { createLogger } from './core'
import { parseArray } from './utils'

const colors = {
  info: green,
  debug: blue,
  warn: yellow,
  error: red,
  scope: cyan,
  time: magenta,
}

export type TransportLevel = LogLevel | 'timer'

export type TransportFn<T extends LogScope = string> = (data: {
  plainLog: string
  time: Date
  msg: any
  level: TransportLevel
  scope?: Keys<T>
  e?: NormalizedError
}) => void
export type NodeLoggerOption<T extends LogScope = string> = LoggerOption & {
  /**
   * log transports
   *
   * built-in: {@link createFileTransport}
   */
  transports?: Arrayable<TransportFn<T>>
}

export function createNodeLoggerConfig(
  transports?: TransportFn<any>[],
  timeFormat: (date: Date) => string = date => date.toLocaleString(),
): [any, any] {
  function getReadableLog(
    time: Date,
    msg: any,
    level: LogLevel,
    scope?: string,
    e?: NormalizedError,
  ): string {
    const _stack = e ? parseStack(e.stack) : e
    const _level = colors[level](level.toUpperCase().padEnd(5))
    const _msg = parseMsg(msg)
    const _time = colors.time(timeFormat(time))
    const _scope = scope ? colors.scope(scope.padEnd(7)) : dim`default`
    return [_time, _level, _scope, _msg + (e ? '\n' + _stack : '')].join(' | ')
  }
  function onNodeLog<T extends LogScope>(msg: any, level: LogLevel, scope?: Keys<T>, err?: unknown): void {
    const time = new Date()
    const e = err ? toNormalizedError(err) : undefined
    const terminalLog = getReadableLog(time, msg, level, scope, e)
    console[level === 'error' ? 'error' : 'log'](terminalLog)
    transports?.forEach(t => t({
      plainLog: ansis.strip(terminalLog),
      msg,
      time,
      level,
      scope,
      e,
    }))
  }

  function onNodeTimer(label: string) {
    const start = Date.now()
    return () => {
      const time = new Date()
      const duration = `${(time.getTime() - start).toFixed(2)}ms`
      const _time = timeFormat(time)
      const logString = `${colors.time(_time)} | ${bold.bgCyan(` ${label} `)} ${duration}`
      console.log(logString)
      transports?.forEach(t => t({
        time,
        plainLog: ansis.strip(logString),
        level: 'timer',
        msg: duration,
        scope: label,
      }))
    }
  }

  return [onNodeLog, onNodeTimer]
}

export function parseMsg(msg: any): string {
  try {
    return typeof msg === 'string'
      ? msg
      : isError(msg)
        ? msg.message
        : JSON.stringify(msg)
  } catch {
    return '' + msg
  }
}

const CWD_REGEXP = new RegExp(process.cwd().replace(/\\/g, '/'), 'i')
const LOCATION_REGEX = /:[^)]+/
const PAREN_REGEX = /\((.*)\)/
const BACK_SLASH_REGEX = /\\/g
export function parseStack(stack: string): string {
  const _s = stack.split(/\r?\n/)
  const _stack = _s
    .slice(1)
    .map(l => l
      .replace('file://', '')
      .replace(' at', blue`@`)
      .replace(BACK_SLASH_REGEX, '/')
      .replace(CWD_REGEXP, '.')
      .replace(LOCATION_REGEX, green)
      .replace(PAREN_REGEX, (_, path) => `(${yellow(path)})`),
    )
  return [_s[0].replace(
    /(.*): (.*)/,
    (_, level, msg) => `${bold.bgRed` ${level} `} ${bold(msg)}`,
  )]
    .concat(_stack)
    .join(EOL)
}

/**
 * Create default node logger
 * @param option logger options, logMode default to `'info'`
 */
export function createNodeLogger<T extends LogScope = string>(option: NodeLoggerOption<T> = {}): Logger<T> {
  const { logMode = 'info', timeFormat, transports } = option
  const list = transports ? parseArray(transports) : undefined

  return createLogger<T>(logMode, ...createNodeLoggerConfig(list, timeFormat))
}
