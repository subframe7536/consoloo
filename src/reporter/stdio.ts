import type { Reporter } from '../type'
import type { MessageFormatter } from './utils'

import { defaultStdFormatter } from './utils'
export interface StdReporterOptions<T extends string> {
  /**
   * Log time format function
   */
  timeFormat?: (date: Date) => string
  /**
   * Custom function to format log message,
   * default to {@link defaultStdFormatter}
   */
  logMessage?: MessageFormatter<T>
}
/**
 * Creates a reporter function that outputs logs to the terminal.
 *
 * Embeded in `createNodeLogger()`
 *
 * @param timeFormat - A function that formats a Date object into a string timestamp,
 * default is `toLocaleString()`
 *
 * @example
 * ```ts
 * const reporter = createStdioReporter((date) => date.toISOString());
 * ```
 */

export function createStdioReporter<T extends string>(
  options: StdReporterOptions<T> = {},
): Reporter<T> {
  const {
    timeFormat = (date: Date) => date.toLocaleString(),
    logMessage = defaultStdFormatter,
  } = options
  return (date, msg, level, scope, e) => {
    const message = logMessage(date, msg, level as any, scope as T, e, timeFormat)
    process[((level === 'error' || level === 'warn') ? 'stderr' : 'stdout')].write(message + '\n')
  }
}
