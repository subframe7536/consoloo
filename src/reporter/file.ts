import type { LogLevel, Reporter } from '../type'

import { appendFileSync, existsSync, mkdirSync, renameSync, statSync } from 'node:fs'
import { join } from 'node:path'

export interface FileReporterOptions<T extends string> {
  /**
   * Log directory path
   * @default './logs'
   */
  logDir?: string
  /**
   * Max file size in `MB` before rotation
   * @default 10
   */
  maxSize?: number
  /**
   * Number of backup files to keep
   * @default 50
   */
  maxBackups?: number
  /**
   * Time format for log entries
   * @default data => data.toLocaleString()
   */
  timeFormat?: (date: Date) => string
  /**
   * Custom function to generate log file name (WITHOUT extension)
   * @default () => 'app'
   * @example
   * ```ts
   * getLogFileName: (level, scope) => `app-${level}-${scope}`
   * ```
   */
  getLogFileName?: (level: LogLevel | 'timer', scope?: T) => string
}

/**
 * Creates a file-based reporter for logging messages to files with rotation capabilities.
 *
 * @param options - {@link FileReporterOptions Configuration} for the file reporter
 *
 * The reporter:
 * - Creates log directory if it doesn't exist
 * - Rotates log files when they exceed maxSize
 * - Maintains separate log files based on level and scope
 *
 * Log file names are generated using `getLogFileName(level, scope) + '.log'`.
 * Rotated files have numbers appended: `filename.1.log`, `filename.2.log`, etc.
 */
export function createFileReporter<T extends string>(
  options: FileReporterOptions<T> = {},
): Reporter<T> {
  const {
    logDir = './logs',
    maxSize = 1,
    maxBackups = 50,
    timeFormat = (date: Date) => date.toLocaleString(),
    getLogFileName = () => 'app',
  } = options

  const fileMap = new Map<string, string>()
  const maxBytes = maxSize << 20 // Convert MB to bytes

  // Ensure log directory exists once at initialization
  if (!existsSync(logDir)) {
    mkdirSync(logDir, { recursive: true })
  }

  function rotateFile(get: (count?: number) => string): void {
    for (let i = maxBackups - 1; i >= 0; i--) {
      const source = i === 0 ? get() : get(i)
      const target = get(i + 1)
      if (existsSync(source)) {
        renameSync(source, target)
      }
    }
  }

  function getPath(level: LogLevel | 'timer', scope?: string): string {
    const name = join(logDir, getLogFileName(level, scope as T))

    const paramKey = `${level}:${scope}`
    let filePath = fileMap.get(paramKey)
    if (!filePath) {
      filePath = name + '.log'
      fileMap.set(paramKey, filePath)
    }
    try {
      const stats = statSync(filePath)
      if (stats.size >= maxBytes) {
        rotateFile(i => i ? name + '.' + i + '.log' : filePath)
      }
    } catch (e) {
      if ((e as { code: string }).code !== 'ENOENT') {
        throw e
      }
    }
    return filePath
  }

  return (date, msg, level, scope, e) => {
    const timestamp = timeFormat(date)
    const filePath = getPath(level, scope)

    const list = [timestamp, level.toUpperCase()]
    if (scope) {
      list.push(scope)
    }
    list.push(msg)
    let logMessage = list.join(' | ')
    if (e) {
      logMessage += `\n${e.message}\n${e.stack}`
    }

    appendFileSync(filePath, logMessage + '\n')
  }
}
