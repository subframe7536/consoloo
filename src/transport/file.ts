import { appendFileSync, existsSync, mkdirSync, renameSync, statSync, writeFileSync } from 'node:fs'
import { dirname } from 'node:path'
import type { Arrayable } from '@subframe7536/type-utils'
import type { TransportFn, TransportLevel } from '../node'
import type { LogScope } from '../type'

type LogFile = {
  /**
   * log file path
   */
  dest: string
  /**
   * log events
   */
  level?: Arrayable<TransportLevel>
  /**
   * max file size
   * @default 1 << 22 (4MB)
   */
  maxSize?: number
}
export type FilePathOrTransportOptions = string | Arrayable<LogFile>
type ParsedOption = Record<TransportLevel, Required<Omit<LogFile, 'level'>>>

function parseOptions(options: FilePathOrTransportOptions): ParsedOption {
  const base = {
    dest: '',
    maxSize: 1 << 22, // 4MB
  }
  const parsedOption: ParsedOption = {
    debug: base,
    info: base,
    warn: base,
    error: base,
    timer: base,
  }
  if (typeof options === 'string') {
    for (const key of Object.keys(parsedOption)) {
      parsedOption[key as TransportLevel].dest = options
    }
    return parsedOption
  }
  if (!Array.isArray(options)) {
    options = [options]
  }
  for (let { dest, level, maxSize } of options) {
    const levels = level
      ? Array.isArray(level)
        ? level
        : [level]
      : ['debug', 'info', 'warn', 'error', 'timer'] as TransportLevel[]
    for (const l of levels) {
      parsedOption[l].dest = dest
      if (maxSize) {
        parsedOption[l].maxSize = maxSize
      }
    }
  }
  return parsedOption
}

function assertFileSize(file: string, maxSize: number) {
  const stat = statSync(file)
  if (stat.size > maxSize) {
    renameSync(file, `${file}.${new Date().getTime()}.bak`)
  }
}
function assertFileExist(option: ParsedOption) {
  for (const { dest } of Object.values(option)) {
    const dir = dirname(dest)
    if (!existsSync(dir)) {
      mkdirSync(dir, { recursive: true })
    }
    if (!existsSync(dest)) {
      writeFileSync(dest, '')
    }
  }
}

export function createFileTransport<T extends LogScope>(
  dest: FilePathOrTransportOptions,
): TransportFn<T> {
  const options = parseOptions(dest)
  assertFileExist(options)
  return ({ plainLog, level }) => {
    const { dest, maxSize } = options[level]
    assertFileSize(dest, maxSize)
    appendFileSync(dest, `${plainLog}\n`)
  }
}
