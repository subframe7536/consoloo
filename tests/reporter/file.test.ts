import { existsSync, mkdirSync, readFileSync, rmSync } from 'node:fs'
import { join } from 'node:path'

import { afterEach, beforeEach, describe, expect, it } from 'bun:test'
import { toNormalizedError } from 'normal-error'

import { createFileReporter } from '../../src/reporter/file'

const LOG_DIR = './test-logs'

describe('File Reporter', () => {
  beforeEach(() => {
    // Clean up and create log directory before each test
    if (existsSync(LOG_DIR)) {
      rmSync(LOG_DIR, { recursive: true, force: true })
    }
    mkdirSync(LOG_DIR, { recursive: true })
  })

  afterEach(() => {
    // Clean up log directory after each test
    if (existsSync(LOG_DIR)) {
      rmSync(LOG_DIR, { recursive: true, force: true })
    }
  })

  it('should log a simple message and match snapshot', () => {
    const reporter = createFileReporter({
      logDir: LOG_DIR,
      timeFormat: () => 'mock-date',
    })
    const date = new Date()
    reporter(date, 'hello world', 'info', 'test')

    // Since file operations are synchronous, we can read the file immediately
    const logFilePath = join(LOG_DIR, 'app.log')
    expect(existsSync(logFilePath)).toBe(true)
    const logContent = readFileSync(logFilePath, 'utf-8')
    expect(logContent).toMatchSnapshot()
  })

  it('should handle error logging and match snapshot', () => {
    const reporter = createFileReporter({
      logDir: LOG_DIR,
      timeFormat: () => 'mock-date',
    })
    const date = new Date()
    const error = new Error('test error')
    error.stack = 'mock stack trace'
    reporter(date, 'an error occurred', 'error', 'app', toNormalizedError(error))

    const logFilePath = join(LOG_DIR, 'app.log')
    expect(existsSync(logFilePath)).toBe(true)
    const logContent = readFileSync(logFilePath, 'utf-8')
    expect(logContent).toMatchSnapshot()
  })

  it('should use custom getLogFileName and match snapshot', () => {
    const reporter = createFileReporter({
      logDir: LOG_DIR,
      timeFormat: () => 'mock-date',
      getLogFileName: (level, scope) => `${scope}-${level}`,
    })
    const date = new Date()
    reporter(date, 'custom file name', 'warn', 'custom-scope')

    const logFilePath = join(LOG_DIR, 'custom-scope-warn.log')
    expect(existsSync(logFilePath)).toBe(true)
    const logContent = readFileSync(logFilePath, 'utf-8')
    expect(logContent).toMatchSnapshot()
  })

  it('should rotate log file when size exceeds maxSize', () => {
    const reporter = createFileReporter({
      logDir: LOG_DIR,
      maxSize: 0.00001, // ~10 bytes, very small to force rotation
      timeFormat: () => 'mock-date',
    })
    const date = new Date()

    // First log, creates app.log
    reporter(date, 'message 1', 'info', 'test')
    const logFilePath = join(LOG_DIR, 'app.log')
    const backupPath1 = join(LOG_DIR, 'app.1.log')
    expect(existsSync(logFilePath)).toBe(true)
    expect(existsSync(backupPath1)).toBe(false)
    const content1 = readFileSync(logFilePath, 'utf-8')
    expect(content1).toContain('message 1')

    // Second log, should trigger rotation
    reporter(date, 'message 2', 'info', 'test')
    expect(existsSync(logFilePath)).toBe(true) // new app.log
    expect(existsSync(backupPath1)).toBe(true) // old app.log becomes app.1.log

    const content2 = readFileSync(logFilePath, 'utf-8')
    const backupContent1 = readFileSync(backupPath1, 'utf-8')

    expect(content2).toContain('message 2')
    expect(backupContent1).toContain('message 1')

    expect(content2).toMatchSnapshot('new_log_after_rotation')
    expect(backupContent1).toMatchSnapshot('rotated_log_content')
  })

  it('should handle multiple backups', () => {
    const reporter = createFileReporter({
      logDir: LOG_DIR,
      maxSize: 0.00001, // ~10 bytes
      maxBackups: 2,
      timeFormat: () => 'mock-date',
      getLogFileName: () => 'test-backup',
    })
    const date = new Date()

    reporter(date, 'message 1', 'info', 'main')
    reporter(date, 'message 2', 'info', 'main')
    reporter(date, 'message 3', 'info', 'main')
    reporter(date, 'message 4', 'info', 'main')

    const logPath = (i?: number): string => join(LOG_DIR, `test-backup${i ? `.${i}` : ''}.log`)

    expect(readFileSync(logPath(), 'utf-8')).toContain('message 4')
    expect(readFileSync(logPath(1), 'utf-8')).toContain('message 3')
    expect(readFileSync(logPath(2), 'utf-8')).toContain('message 2')
    expect(existsSync(logPath(3))).toBe(false)
  })
})
