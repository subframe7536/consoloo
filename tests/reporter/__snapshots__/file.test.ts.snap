// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`File Reporter should log a simple message and match snapshot 1`] = `
"mock-date | INFO | test | hello world
"
`;

exports[`File Reporter should handle error logging and match snapshot 1`] = `
"mock-date | ERROR | app | an error occurred
test error
mock stack trace
"
`;

exports[`File Reporter should use custom getLogFileName and match snapshot 1`] = `
"mock-date | WARN | custom-scope | custom file name
"
`;

exports[`File Reporter should rotate log file when size exceeds maxSize: new_log_after_rotation 1`] = `
"mock-date | INFO | test | message 2
"
`;

exports[`File Reporter should rotate log file when size exceeds maxSize: rotated_log_content 1`] = `
"mock-date | INFO | test | message 1
"
`;
